File: Java/src/session1/td/ZeroRec.java
package session1.td;

public class ZeroRec extends Zero{

	public ZeroRec() {
		super();
	}
	
	@Override
	public Nat somme(Nat n) {
		return n;
	}
	
	@Override
	public Nat produit(Nat n) {
		System.out.print("PasseZero");
		return this.creerZero();
	}
	
	@Override
	public boolean equals(Object o) {
		return ((Nat)o).val() == 0;
	}
	
	@Override
	public Nat div(Nat x) {
		return this.creerZero();
	}
	
	@Override
	public Nat modulo(Nat x) {
		return super.creerZero();
	}

}

File: Typescript/ZeroRec.ts
import { Zero } from "../Zeroion1.td/Zero";
import { Nat } from "./session1.td/Natt";

export class ZeroRec extends Zero{
    constructor(){
        super();
    }

    somme(n:Nat):Nat{
        return n;
    }

    produit(n:Nat):Nat{
        return this.creerZero();
    }

    equals(o:Object):boolean{
        return (o as Nat).val()==0;
    }

    div(n:Nat):Nat{
        if(n.val()!=0){
            return this.creerZero();
        } else {
            throw new Error("Division par zero");
        }
    }
    modulo(n:Nat):Nat{
        if(n.val()!=0){
            return this.creerZero();
        } else {
            throw new Error("Division par zero");
        }
    }
}